crates/sui-core/src/narwhal_manager/mod.rs

```diff
@@ -54,128 +33,127 @@ pub struct NarwhalConfiguration<TxValidator: TransactionValidator> {
     pub registry_service: RegistryService,
 }
 
-pub async fn run_narwhal_manager<State, TxValidator>(
-    config: NarwhalConfiguration<TxValidator>,
-    mut tr_start: Receiver<NarwhalStartMessage<State>>,
-    mut tr_stop: Receiver<()>,
-) where
-    State: ExecutionState + Send + Sync + 'static,
-    TxValidator: TransactionValidator,
-{
-    // Create the Narwhal Primary with configuration
-    let primary_node = PrimaryNode::new(
-        config.parameters.clone(),
-        true,
-        config.registry_service.clone(),
-    );
-
-    // Create Narwhal Workers with configuration
-    let worker_nodes = WorkerNodes::new(config.registry_service.clone(), config.parameters.clone());
-
-    loop {
-        // Copy the config for this iteration of the loop
-        let mut id_keypair_copy = Vec::new();
-        for (id, keypair) in &config.worker_ids_and_keypairs {
-            id_keypair_copy.push((*id, keypair.copy()));
+pub struct NarwhalManager<TxValidator> {
+    storage_base_path: PathBuf,
+    primary_keypair: bls12381::min_sig::BLS12381KeyPair,
+    network_keypair: NetworkKeyPair,
+    worker_ids_and_keypairs: Vec<(WorkerId, NetworkKeyPair)>,
+    primary_node: PrimaryNode,
+    worker_nodes: WorkerNodes,
+    tx_validator: TxValidator,
+    running: Mutex<bool>,
+}
+
+impl<TxValidator: TransactionValidator> NarwhalManager<TxValidator> {
+    pub fn new(config: NarwhalConfiguration<TxValidator>) -> Self {
+        // Create the Narwhal Primary with configuration
+        let primary_node = PrimaryNode::new(
+            config.parameters.clone(),
+            true,
+            config.registry_service.clone(),
+        );
+
+        // Create Narwhal Workers with configuration
+        let worker_nodes =
+            WorkerNodes::new(config.registry_service.clone(), config.parameters.clone());
+
+        Self {
+            primary_node,
+            worker_nodes,
+            tx_validator: config.tx_validator,
+            primary_keypair: config.primary_keypair,
+            network_keypair: config.network_keypair,
+            worker_ids_and_keypairs: config.worker_ids_and_keypairs,
+            storage_base_path: config.storage_base_path,
+            running: Mutex::new(false),
+        }
+    }
+
+    // Starts the Narwhal (primary & worker(s)) - if not already running.
+    pub async fn start<State>(
+        &self,
+        committee: Arc<Committee>,
+        shared_worker_cache: SharedWorkerCache,
+        execution_state: Arc<State>,
```

Comment:
Since we now pass in the state as a start parameter, this prevents the need for generics on SuiNode 👍